NetSDK/Services/SegmentFetcher/Classes/SelfRefreshingSegmentFetcher.cs

```diff
@@ -1,47 +1,84 @@
+using System.ComponentModel;
 using System.Linq;
 using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
 
 namespace Splitio.Services.SegmentFetcher.Classes
 {
-    public class SelfRefreshingSegmentFetcher : ISegmentFetcher
+    public class SelfRefreshingSegmentFetcher : SegmentFetcher
     {
         private static readonly ILog Log = LogManager.GetLogger(typeof(SelfRefreshingSegmentFetcher));
+        
         private readonly ISegmentChangeFetcher segmentChangeFetcher;
-        private ConcurrentDictionary<string, SelfRefreshingSegment> segments;
+        private ConcurrentDictionary<string, SelfRefreshingSegment> segmentsThreads;
+        private SegmentTaskWorker worker;
         private SdkReadinessGates gates;
         private int interval;
+        private CancellationTokenSource cancelTokenSource = new CancellationTokenSource(); 
 
-        public SelfRefreshingSegmentFetcher(ISegmentChangeFetcher segmentChangeFetcher, SdkReadinessGates gates, int interval, ConcurrentDictionary<string, SelfRefreshingSegment> segments = null)
+        public SelfRefreshingSegmentFetcher(ISegmentChangeFetcher segmentChangeFetcher, SdkReadinessGates gates, int interval, ISegmentCache segmentsCache, int numberOfParallelSegments):base(segmentsCache)
         {
             this.segmentChangeFetcher = segmentChangeFetcher;
-            this.segments = segments ?? new ConcurrentDictionary<string, SelfRefreshingSegment>();
+            this.segmentsThreads = new ConcurrentDictionary<string, SelfRefreshingSegment>();
+            worker = new SegmentTaskWorker(numberOfParallelSegments); 
             this.interval = interval;
             this.gates = gates;
+            StartWorker();
         }
 
-        public Segment Fetch(string name)
+        public void Stop()
+        {
+            cancelTokenSource.Cancel();
+        }
+
+        private void StartWorker()
+        {
+            Task workerTask = Task.Factory.StartNew(
+                () => worker.ExecuteTasks(), 
+                cancelTokenSource.Token);
+        }
+
+        public void StartScheduler()
+        {
+            //Delay first execution until expected time has passed
+            Thread.Sleep(interval * 1000);
+            Task schedulerTask = PeriodicTaskFactory.Start(
+                    () => AddSegmentsToQueue(),
+                    intervalInMilliseconds: interval * 1000,
+                    cancelToken: cancelTokenSource.Token);
+        }
+
+        public override void InitializeSegment(string name)
         {
             SelfRefreshingSegment segment;
-            segments.TryGetValue(name, out segment);
-            if (segment != null)
+            segmentsThreads.TryGetValue(name, out segment);
+            if (segment == null)
             {
-                return segment;
+                segment = new SelfRefreshingSegment(name, segmentChangeFetcher, gates, segmentCache);
+                gates.RegisterSegment(name);
```

Comment:
It starts when I call 
 SegmentTaskQueue.segmentsQueue.TryAdd(segment); 

and the worker dequeues it